feat: Container Jump Improvements and Code for tracking Container Jumps

rahulramesha opened the pull request:
## Description
This PR is to optimise container Jumps and add code to track user behaviour on container jumps.
Changes:
* moving Container jump thresholds to absolute values
* adding additional check for container jump that the mouse pointer has to be on the container before jumping
* Additional code to track container jumps from main Canvas.

Fixes #12690
Fixes #11487

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Manual and jest

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes



## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>


    // Code coverage diff between base branch:release and head branch: containerJumpOptimization 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.34 **(0)** | 37.72 **(-0.02)** | 35.97 **(0.02)** | 56.56 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/PositionedContainer.tsx | 100 **(0)** | 82.54 **(-1.33)** | 100 **(0)** | 100 **(0)**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/hooks/ContainerJumpMetric.ts** | **88.89** | **100** | **75** | **88.89**
 :red_circle: | app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts | 89.44 **(-5.49)** | 76.47 **(-10.74)** | 90.63 **(0.63)** | 89.26 **(-6.09)**
 :red_circle: | app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragging.ts | 83.64 **(-1.25)** | 67.05 **(-0.63)** | 84.38 **(0)** | 84.19 **(-1.43)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 88.66 **(0.62)** | 76 **(0)** | 85.71 **(1.09)** | 85.33 **(0.82)**
 :red_circle: | app/client/src/utils/hooks/useReflow.ts | 81.69 **(-9.49)** | 60.61 **(-16.06)** | 100 **(0)** | 86.36 **(-5.7)**
 :red_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 15.32 **(-0.55)** | 0 **(0)** | 4.55 **(0)** | 15.93 **(-0.59)**</details>

rahulramesha commented:
/ok-to-test sha=e0ceade

riodeuno commented on app/client/src/utils/WidgetPropsUtils.test.tsx:
```diff
@@ -70,7 +71,24 @@ describe("WidgetProps tests", () => {
+      parentRowSpace: 10,
+      maxGridColumns: 64,
+    };
+    const mouseEvent = ({
```
Please add another test case for really small offset values (1 or 2)

rahulramesha replied:
Added a test, for testing low Offset numbers, that generates negative values of mouse Positions, which is expected.

Also, today I learnt, javascript has `-0`.

riodeuno commented on app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts:
```diff
@@ -104,6 +111,52 @@ export const useBlocksToBeDraggedOnCanvas = ({
   const { updateWidget } = useContext(EditorContext);
 
   const allWidgets = useSelector(getWidgets);
+  //This method is called whenever a there is a canvas change
```
canvas (DropTarget)

riodeuno replied:
Also, what does this method do?

riodeuno commented on app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts:
```diff
@@ -104,6 +111,52 @@ export const useBlocksToBeDraggedOnCanvas = ({
   const allWidgets = useSelector(getWidgets);
+  //This method is called whenever a there is a canvas change
+  const logContainerJump = (
+    canvasId: string,
```
Let's rename this to make sense of what this value containers. `dropTargetWidgetId`

riodeuno commented on app/client/src/utils/WidgetPropsUtils.tsx:
```diff
@@ -96,6 +102,28 @@ export const getDropZoneOffsets = (
+  e: MouseEvent,
+  gridProps: GridProps,
+) => {
+  const mouseTop = Math.round(
```
Please verify if Math.floor would be appropriate here.

riodeuno reviewed (changes requested):
A few changes requested. The rest looks good to me.

rahulramesha commented:
@riodeuno , addressed most of the review comments, `useReducer` did not work in this situation because, placing it inside the Instance causes it to re-render whenever the value changes. Decided to convert it into a class instead for better readability.

rahulramesha commented:
/ok-to-test sha=161c6c5

vivekverma2312 commented:
@rahulramesha Tested and working fine.
